internal/pkg/nats/client.go

```diff
@@ -0,0 +1,180 @@
+	return NewClientWithConnector(cfg, func(broker types.HostInfo, option ...nats.Option) (Connection, error) {
+		url := nats.DefaultURL
+		if !broker.IsHostInfoEmpty() {
+			url = broker.GetHostURL()
+		}
+
+		nc, err := nats.Connect(url, option...)
+
+		if err != nil {
+			return nil, err
+		}
+
+		return nc, nil
+	})
+}
+
+// NewClientWithConnector creates a new client that uses the specified function to establish pub/sub connections
+func NewClientWithConnector(cfg types.MessageBusConfig, connector ConnectNats) (*Client, error) {
+	if connector == nil {
+		return nil, fmt.Errorf("connector is required")
+	}
+
+	var m MarshallerUnmarshaller = &edgexMarshaller{} // default
+
+	switch strings.ToLower(cfg.Optional[format]) {
+	case "nats":
+		m = &natsMarshaller{}
+	case "json":
+		m = &jsonMarshaller{}
+	case "cbor":
+		m = &cborMarshaller{}
+	}
+
+	return &Client{config: cfg, connect: connector, m: m}, nil
+}
+
+// Client provides NATS MessageBus implementations per the underlying connection
+type Client struct {
+	connect ConnectNats
+	subConn Connection
+	pubConn Connection
+	m       MarshallerUnmarshaller
+	config  types.MessageBusConfig
+}
+
+// Connect establishes the connections to publish and subscribe hosts
+func (c *Client) Connect() error {
+	if c.connect == nil {
+		return fmt.Errorf("connection function not specified")
+	}
+
+	subConnectOpt, err := buildSubscriberConnectOpt(c.config.SubscribeHost, c.config.Optional)
+
+	if err != nil {
+		return err
+	}
+
+	subConn, err := c.connect(c.config.SubscribeHost, subConnectOpt...)
+
+	if err != nil {
+		return err
+	} else {
+		c.subConn = subConn
+	}
+
+	if !c.config.PublishHost.IsHostInfoEmpty() {
+		pubConnectOpt, err := buildPublisherConnectOpt(c.config.SubscribeHost, c.config.Optional)
+
+		if err != nil {
+			return err
+		}
+
+		pubConn, err := c.connect(c.config.PublishHost, pubConnectOpt...)
+
+		if err != nil {
+			return err
+		} else {
+			c.pubConn = pubConn
+		}
+	}
+
+	return nil
+}
+
+// Publish publishes EdgeX messages to NATS
+func (c *Client) Publish(message types.MessageEnvelope, topic string) error {
+	if c.pubConn == nil {
+		return fmt.Errorf("cannot publish with disconnected client")
+	}
+
+	if topic == "" {
+		return fmt.Errorf("cannot publish to empty topic")
+	}
+
+	msg, err := c.m.Marshal(message, topic)
+
+	if err != nil {
+		return err
+	}
+
+	return c.pubConn.PublishMsg(msg)
+}
+
+// Subscribe establishes NATS subscriptions for the given topics
+func (c *Client) Subscribe(topics []types.TopicChannel, messageErrors chan error) error {
+	if c.subConn == nil {
+		return fmt.Errorf("cannot subscribe with disconnected client")
+	}
+
+	for _, tc := range topics {
+		s := topicToSubject(tc.Topic)
+
+		_, err := c.subConn.QueueSubscribe(s, c.config.Optional[QueueGroup], func(msg *nats.Msg) {
+			env := types.MessageEnvelope{}
+			err := c.m.Unmarshal(msg, &env)
+			if err != nil {
+				messageErrors <- err
+			} else {
+				tc.Messages <- env
```

Comment:
Need to set ReceivedTopic on MessageEnvelope. It needs to be the actual topic that was publish, not the subscribe topic with wild cards. How do we get that from NATS?

Reply:
The unmarshalers handle getting this directly from the NATS message itself.